### PR TITLE
gh-112636: remove check for float subclasses without nb_float

### PR DESCRIPTION
**Pull request by skirpichev**
This condition could be triggered only in very special scenario, if users code explicitly override inherited ``nb_float`` value *and* set it to ``NULL``:
```c
/* nb_float will be inherited anyway if it's set to NULL here,
   so this is essentially redundant.  */
static PyNumberMethods xxx_as_number = {
    .nb_float = NULL,
};

static PyObject*
break_it(PyObject *o, PyObject *Py_UNUSED(ignored))
{
    Py_TYPE(o)->tp_as_number->nb_float = NULL;
    Py_RETURN_NONE;
}

static PyMethodDef xxx_methods[] = {
    {"break_it", (PyCFunction) break_it, METH_NOARGS, NULL},
    {NULL},
};

PyTypeObject XXX_Type = {
    PyVarObject_HEAD_INIT(NULL, 0)
    .tp_name = "xxx",
    .tp_base = &PyFloat_Type,
    .tp_as_number = &xxx_as_number,
    .tp_methods = xxx_methods,
};
```
```diff
diff --git a/Objects/abstract.c b/Objects/abstract.c
index 7cca81464c..41216eeb9b 100644
--- a/Objects/abstract.c
+++ b/Objects/abstract.c
@@ -1640,6 +1640,7 @@ PyNumber_Float(PyObject *o)
 
     /* A float subclass with nb_float == NULL */
     if (PyFloat_Check(o)) {
+        printf("XXX\n");
         return PyFloat_FromDouble(PyFloat_AS_DOUBLE(o));
     }
     return PyFloat_FromString(o);

```
```pycon
>>> a = xxx(3.14)
>>> float(a)  # nb_float == NULL is not triggered, due to inheritance
3.14
>>> a.break_it()
>>> float(a)  # and only NOW this branch will work...
XXX
3.14
```
-----------------------------

* Issue: gh-112636

### CODE REVIEW

**Comment by skirpichev** [comment]
Probably, this could skip news.

**Review by serhiy-storchaka — DISMISSED** [review]

**Comment by skirpichev** [comment]
@serhiy-storchaka, could you please comment what's wrong with the pr on your opinion?  Apparently, if removed block should be kept - then we must restore similar block in PyNumber_Long(), removed in 31a655411.

**Comment by skirpichev** [comment]
CC @vstinner, does it make sense for you?

**Comment by vstinner** [comment]
```
>>> a.break_it()
>>> float(a)  # and only NOW this branch will work...
XXX
3.14
```

I don't understand the example. float(a) does now return XXX? Or does it print XXX? Which part of the code prints XXX?

**Comment by skirpichev** [comment]
> I don't understand the example. float(a) does now return XXX?

Ah, that was just a debug print from the removed (by this pr) code.  I've added a diff to description.

The point is that to trigger that code - you must explicitly set nb_float field in the derived class to NULL.  But in this way we can break everything!  Can't you set nb_add to NULL in a float subclass?  Easy!  But CPython has enough code, which assuming that float subclasses have working arithmetic dunder methods.

**Comment by vstinner** [comment]
What is the behavior on this `nb_float=NULL` type with your change? Does `return PyFloat_FromString(o)` fail?

**Comment by skirpichev** [comment]
> What is the behavior on this nb_float=NULL type with your change?

It fails with TypeError (after you run first ``break_it()``).

> Does return PyFloat_FromString(o) fail?

Yes, because ``o`` is not a subclass of str/bytes/etc.

**Review by vstinner — COMMENTED** [review]
This change is a backward incompatible change. You should document it in a NEWS entry.

**Comment by serhiy-storchaka** [comment]
Sorry, but I now think that we should rather return a fallback for integers.

**Comment by skirpichev** [comment]
> This change is a backward incompatible change. You should document it in a NEWS entry.

Ok, I did.  News entry was missed just as in 31a655411 (dropped similar check in PyNumber_Long).

> I now think that we should rather return a fallback for integers.

But why?!  Should we check on same ground that an int subtype has nb_add field set?

**Comment by serhiy-storchaka** [comment]
Because `PyFloat_AsDouble()` never fails for instances of `float` subclasses. It does not use `nb_float` in that case.

**Comment by skirpichev** [comment]
In same way we can recover in case when float subtype sets e.g. nb_negative to ``NULL``.  I think such subtypes are broken and it's better to fail quickly.

File: Misc/NEWS.d/next/C_API/2024-10-01-14-53-55.gh-issue-112636.t4WJjC.rst [hunk 1]
```diff
@@ -0,0 +1,4 @@
+:c:func:`PyNumber_Float` doesn't fallback anymore to
+``PyFloat_FromDouble(PyFloat_AS_DOUBLE)`` if a :class:`float` subtype reset
+:c:member:`~PyNumberMethods.nb_float` to ``NULL``.  A :exc:`TypeError`
+exception raised instead.
```
Comment (vstinner) [inline Misc/NEWS.d/next/C_API/2024-10-01-14-53-55.gh-issue-112636.t4WJjC.rst]:
```suggestion
exception is raised instead.
```

File: Misc/NEWS.d/next/C_API/2024-10-01-14-53-55.gh-issue-112636.t4WJjC.rst [hunk 2]
```diff
@@ -0,0 +1,4 @@
+:c:func:`PyNumber_Float` doesn't fallback anymore to
+``PyFloat_FromDouble(PyFloat_AS_DOUBLE)`` if a :class:`float` subtype reset
```
Comment (vstinner) [inline Misc/NEWS.d/next/C_API/2024-10-01-14-53-55.gh-issue-112636.t4WJjC.rst]:
```suggestion
``PyFloat_FromDouble(PyFloat_AS_DOUBLE(obj))`` if a :class:`float` subtype resets
```

**Comment by skirpichev** [comment]
> now think that we should rather return a fallback for integers.

There is a difference, if nb_int is set to ``NULL``, but nb_index is not broken (at least not ``NULL``) - then we got an equivalent fallback with ``_PyLong_Copy``, see:
https://github.com/python/cpython/blob/31516c98dd7097047ba10da8dcf728c3d580f3d6/Objects/abstract.c#L1403-L1405
and
https://github.com/python/cpython/blob/31516c98dd7097047ba10da8dcf728c3d580f3d6/Objects/abstract.c#L1446-L1448

I still think we shouldn't keep workarounds for broken subtypes.  But if so, at least the nb_int workaround shouldn't be restored.

**Comment by skirpichev** [comment]
Ok, I'm closing this.

**Pull request closed without merging**